### PR TITLE
Implement MultiAssetProxy

### PR DESCRIPTION
**Pull request by abandeali1**
## Description
Implements https://github.com/0xProject/ZEIPs/issues/23

TODO: add combinatorial tests
TODO: use custom ABI encoder and update `assetDataUtils`
TODO: determine if we can safely remove sanity checks

## Checklist:

*   [ ] Prefix PR title with `[WIP]` if necessary.
*   [ ] Prefix PR title with bracketed package name(s) corresponding to the changed package(s). For example: `[sol-cov] Fixed bug`.
*   [ ] Add tests to cover changes as needed.
*   [ ] Update documentation as needed.
*   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                    }
+
+                    // Write `totalAmount` to memory
+                    mstore(100, totalAmount)
+
+                    // Load offset to `nestedAssetData[i]`
+                    let nestedAssetDataElementOffset := calldataload(add(nestedAssetDataContentsStart, i))
+
+                    // Load length of `nestedAssetData[i]`
+                    let nestedAssetDataElementContentsStart := add(nestedAssetDataElementOffset, 264)
```
Comment (Recmo) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: `:= add(assetDataOffset, add(nestedAssetDataElementOffset, 4 + 4 + 32))`. (I think)

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                        mstore(96, 0x4544000000000000000000000000000000000000000000000000000000000000)
+                        revert(0, 100)
+                    }
+
+                    // Load AssetProxy id
+                    let assetProxyId := and(calldataload(nestedAssetDataElementContentsStart), 0xffffffff00000000000000000000000000000000000000000000000000000000)
+
+                    // To lookup a value in a mapping, we load from the storage location keccak256(k, p),
+                    // where k is the key left padded to 32 bytes and p is the storage slot
+                    mstore(132, assetProxyId)
```
Comment (Recmo) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: 132 is where `nestedAssetDataElementLen` should be stored for our outgoing calldata. `assetProxyId` would starting from 164.

Edit: doesn't matter, we overwrite anyway in the big calldatacopy. It does seem worth re-using these 4 bytes.

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x0000001a41535345545f50524f58595f444f45535f4e4f545f45584953540000)
+                        mstore(96, 0)
+                        revert(0, 100)
+                    }
+    
+                    // Copy `nestedAssetData[i]` from calldata to memory
+                    calldatacopy(
+                        132,                                // memory slot after `amounts[i]`
+                        nestedAssetDataElementOffset,       // location of `nestedAssetData[i]` in calldata
```
Comment (Recmo) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: This does not take offset from outer assetData into account, nor the selectors. Should be `add(assetDataOffset, add(nestedAssetDataElementOffset, 4 + 4))` or `sub(nestedAssetDataElementContentsStart, 32)`.

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+                    // call `assetProxy.transferFrom`
+                    let success := call(
+                        gas,                                    // forward all gas
+                        assetProxy,                             // call address of asset proxy
+                        0,                                      // don't send any ETH
+                        0,                                      // pointer to start of input
+                        add(164, nestedAssetDataElementLen),    // length of input  
+                        100,                                    // write output over memory that won't be reused
+                        512                                     // reserve 512 bytes for output
```
Comment (Recmo) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: This assumes or proxies will never return more then 512 bytes. Reasonable, but why not use returndatacopy and leave both of these zero?

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                    )
+
+                    if iszero(success) {
+                        revert(100, returndatasize())
```
Comment (Recmo) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]:
```
returndatacopy(0, 0, returndatasize())
revert(0, returndatasize())
```

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                    let assetProxyId := and(
+                        calldataload(nestedAssetDataElementContentsStart),
+                        0xffffffff00000000000000000000000000000000000000000000000000000000
+                    )
+
+                    // To lookup a value in a mapping, we load from the storage location keccak256(k, p),
+                    // where k is the key left padded to 32 bytes and p is the storage slot
+                    mstore(132, assetProxyId)
+                    mstore(164, assetProxies_slot)
+                    let assetProxy := sload(keccak256(132, 64))
```
Comment (hysz) [inline packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: One possible optimization: If `assetProxyId` is the same as the previous iteration then we could skip this check. Clients could then optimize their calldata by grouping-by-proxy.

File: packages/contracts/test/asset_proxy/proxies.ts [hunk 7]
```diff
@@ -613,16 +676,569 @@ describe('Asset Transfer Proxies', () => {
...
+                    from: owner,
+                    to: multiAssetProxy.address,
+                    value: constants.ZERO_AMOUNT,
+                    data: undefinedSelector,
+                }),
+            );
+        });
+        it('should have an id of 0x94cfcdd7', async () => {
+            const proxyId = await multiAssetProxy.getProxyId.callAsync();
+            const expectedProxyId = '0x94cfcdd7';
```
Comment (hysz) [inline packages/contracts/test/asset_proxy/proxies.ts]: Could be good to add a note about how this was derived.

**Comment by hysz** [comment]
🔥 🔥 🔥

File: packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+                    // call `assetProxy.transferFrom`
+                    let success := call(
+                        gas,                                    // forward all gas
+                        assetProxy,                             // call address of asset proxy
+                        0,                                      // don't send any ETH
+                        0,                                      // pointer to start of input
+                        add(164, nestedAssetDataElementLen),    // length of input  
+                        0,                                      // write output over memory that won't be reused
+                        0                                       // reserve 512 bytes for output
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply packages/contracts/contracts/protocol/AssetProxy/MultiAssetProxy.sol]: Whoops, outdated comment

**Pull request merged**